### PR TITLE
Signal Generation Restructure and addition of high level methods

### PR DESCRIPTION
**Pull request by qthompso**
## Proposed changes

This PR contains a fundamental change to how the signal generator (previously signal source) class is structured.
Along with this, there is a the introduction of many more high level functions to help with generating functions.
Please view the signal_generators.md documentation file for an in depth view of each of these methods.

### Added

- Added the constraint ranges for all signal generators
- Added drivers for AWG and AFG channels
- Added a property named `source_channel` in AWG's and AFG's.
- Added drivers for internal AFG in TekScopes.
- Added a property named `internal_afg` in TekScope.
- Added implementation of `generate_function` for all AWG models.
- Added two burst functions to SignalGeneratorMixin: one to set up burst and one to generate the burst by forcing trigger.
  - NOTE: Only the AFG's and internal AFG have these functions implemented.
- Added `OutputSignalPath` enum attribute in AWG's representing output signal path options.
- Added two functions for loading waveform set files in the AWG70k's and AWG5200: one for loading a waveform set file and another for loading a specific waveform from a waveform set file.
- Added `sample_waveform_set_file` attribute in AWG70k's and AWG5200 to define the default waveform set file.

### Changed

- Changed the term "signal source" to "signal generator".
  - NOTE: <span style="color:red">BREAKING CHANGE</span>. All uses of this term are changed. Import paths now use signal_generator instead of signal_source.
- Changed the function name of `generate_waveform()` to `generate_function()`.
  - NOTE: <span style="color:red">BREAKING CHANGE</span>. `generate_waveform()` only exists on AWGs now, however the functionality is entirely changed.
- Changed the `generate_function()` function by removing burst functionality.
  - NOTE: <span style="color:red">BREAKING CHANGE</span>. Any use of burst now must use `setup_burst()` and `generate_burst()` instead.
- Updated AWG's such that the `family_base_class` is at the model level.

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Functionality update (non-breaking change which updates or changes existing functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] CI/CD update (an update to the CI/CD workflows, scripts, and/or configurations)
- [x] Documentation update (an update to enhance the user experience when reading through the docs)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have followed the guidelines in the [CONTRIBUTING](https://github.com/tektronix/tm_devices/blob/main/CONTRIBUTING.md) document
- [x] I have signed the CLA
- [x] I have checked to ensure there aren't other open [Pull Requests](https://github.com/tektronix/tm_devices/pulls) for the same update/change
- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] Basic linting passes locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have added necessary documentation (if appropriate)
- [x] I have updated the Changelog with a brief description of my changes

### CODE REVIEW

File: docs/advanced/signal_generators.md [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+
+Arbitrary Waveform Generators (\[AWGs\](default: AWG)) enforce one cycle per sample, allowing the output to be the same shape regardless of clock rate.
+The number of samples that occurs during a second is referred to as a sample per second (S/s), a unit which determines the frequency of the waveform.
+With low frequency functions, \[AWGs\](default: AWG) are functionally identical to \[AFGs\](default: AFG), besides offering more constrained amplitudes and offsets.
+
+---
+
+### Class Structure
+
+```{mermaid}
```
Comment (nfelt14) [inline docs/advanced/signal_generators.md:26]:
```suggestion
```mermaid
```

File: docs/advanced/signal_generators.md [hunk 2]
```diff
@@ -0,0 +1,638 @@
...
+
+---
+
+## Function Generation
+
+A function is a limited set of common waveforms that are provided by default through the instrument.
+The simplicity of these waveforms allows for output parameters like waveform length and sample rate to be
+abstracted away. Frequency replaces these in order to provide signals which are easy to quantify and manipulate.
+
+Arbitrary Function Generators (\[AFGs\](default: AFG)) utilize a phase increment process and a data lookup table to provide variable frequencies.
```
Comment (nfelt14) [inline docs/advanced/signal_generators.md:13]:
```suggestion
Arbitrary Function Generators ([AFGs](default: AFG)) utilize a phase increment process and a data lookup table to provide variable frequencies.
```

Comment (nfelt14) [reply docs/advanced/signal_generators.md:13]: All the square brackets seem to have been escaped. That will render them all useless.

File: docs/advanced/signal_generators.md (hunk 1 above)
Comment (nfelt14) [reply docs/advanced/signal_generators.md:26]: This needs to be propagated to all mermaid diagrams

File: docs/advanced/signal_generators.md [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+This includes:
+
+- Features unique to each which should be kept in mind.
+- Constraints for each waveform parameter.
+
+---
+
+### TekScope Internal Arbitrary Function Generators
+
+```{autoclasstree} tm_devices.drivers.pi.scopes.tekscope
```
Comment (nfelt14) [inline docs/advanced/signal_generators.md:156]: The new syntax would be this:

```
{{ auto_class_diagram('tm_devices.drivers.pi.scopes.tekscope', full=True, namespace='tm_devices.drivers.pi.scopes.tekscope') }}
```

There is no alignment or alt text now.

Comment (nfelt14) [reply docs/advanced/signal_generators.md:156]: If we need to add it, we will need to update the macro in the docs folder.

File: docs/advanced/signal_generators.md [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+```{note}
+[IAFGs](default: IAFG) contain no waveform list, editable memory or user defined waveforms. This means arbitrary waveforms
+must be loaded from the hard drive.
+```
+
+```{note}
+Some functions, like `SINC`, `GAUSSIAN`, `LORENTZ`, `ERISE`, `EDECAY` and `HAVERSINE` already have an inbuilt offset.
+```
+
+```{note}
```
Comment (nfelt14) [inline docs/advanced/signal_generators.md:193]: All of these notes/hints/warnings/cautions are using Sphinx syntax. They need to be updated to use mkdocs syntax.

File: docs/advanced/signal_generators.md [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+[`MSO4B`][tm_devices.drivers.pi.scopes.tekscope.mso4b.MSO4B]
+[`MSO5`][tm_devices.drivers.pi.scopes.tekscope.mso5.MSO5]
+[`MSO5LP`][tm_devices.drivers.pi.scopes.tekscope.mso5lp.MSO5LP]
+[`MSO6`][tm_devices.drivers.pi.scopes.tekscope.mso6.MSO6]
+[`MSO6B`][tm_devices.drivers.pi.scopes.tekscope.mso6b.MSO6B]
+[`LPD6`][tm_devices.drivers.pi.scopes.tekscope.lpd6.LPD6]
+
+Sample rates are 250.0MS/s for `ARBITRARY` waveforms.
+
+```{table} IAFG Constraints
```
Comment (nfelt14) [inline docs/advanced/signal_generators.md:218]: All of these seem to still be sphinx syntax.

**Review by nfelt14 — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 6]
```diff
@@ -16,7 +16,31 @@ Valid subsections within a version are:
 
 ## Unreleased
 
-Things to be included in the next release go here.
```
Comment (nfelt14) [inline CHANGELOG.md:19]: This line needs to stay here.

File: CHANGELOG.md [hunk 7]
```diff
@@ -16,7 +16,31 @@ Valid subsections within a version are:
...
+---
```
Comment (nfelt14) [inline CHANGELOG.md:19]: This should not be added.

File: docs/known_words.txt [hunk 8]
```diff
@@ -5,7 +5,9 @@ alias_usage
 api
 artifact
 artifacts
+autoclasstree
```
[the remark this answers is not in the document]
Comment (nfelt14) [reply docs/known_words.txt:8]: This needs to be removed

File: docs/known_words.txt [hunk 9]
```diff
@@ -47,6 +51,7 @@ licensor
 linter
 linters
 linting
+literalinclude
```
Comment (nfelt14) [inline docs/known_words.txt:54]: Remove this

**Review by nfelt14 — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 10]
```diff
@@ -161,49 +185,6 @@ Things to be included in the next release go here.
 
 ---
 
-## v1.2.2 (2024-04-03)
```
Comment (nfelt14) [inline CHANGELOG.md:164]: Why was this deleted?

File: CHANGELOG.md [hunk 11]
```diff
@@ -149,7 +173,7 @@ Things to be included in the next release go here.
 
 ---
 
-## v1.2.3 (2024-04-03)
+## v1.2.2 (2024-04-03)
```
Comment (nfelt14) [inline CHANGELOG.md:176]: Why was this changed?

File: src/tm_devices/helpers/constants_and_dataclasses.py [hunk 12]
```diff
@@ -705,3 +705,14 @@ def __str__(self) -> str:
...
+LOAD_IMPEDANCE_LOOKUP: Final[Mapping[Union[float, str], LoadImpedanceAFG]] = MappingProxyType(
+    {
+        9.97e37: LoadImpedanceAFG.HIGHZ,
+        1.0e6: LoadImpedanceAFG.HIGHZ,
+        "HIGHZ": LoadImpedanceAFG.HIGHZ,
+        50.0: LoadImpedanceAFG.FIFTY,
+        "FIFTY": LoadImpedanceAFG.FIFTY,
+    }
+)
+"""Conversions of literal values to Enum values."""
```
Comment (nfelt14) [inline src/tm_devices/helpers/constants_and_dataclasses.py:718]:
```suggestion
"""Conversions of literal values representing impedances to Enum values representing impedances."""
```

**Review by nfelt14 — CHANGES_REQUESTED** [review]

**Review by nfelt14 — DISMISSED** [review]

File: docs/advanced/signal_generators.md [hunk 13]
```diff
@@ -0,0 +1,617 @@
...
+!!! note
+    [`SignalGeneratorMixin`][tm_devices.driver_mixins.signal_generator_mixin.SignalGeneratorMixin]
+    only contains abstract methods; defining the class by itself and calling
+    methods in it will only raise `NotImplemented` errors.
+
+```mermaid
+classDiagram
+    direction LR
+
+    Tekscope <|-- InternalAFGChannel
```
[the remark this answers is not in the document]
Comment (nfelt14) [reply docs/advanced/signal_generators.md:55]: @qthompso, can you double check this diagram?

[the remark this answers is not in the document]
Comment (qthompso) [reply docs/advanced/signal_generators.md:55]: It follows the same direction as the inheritance diagrams in the Architectural Overview page

[the remark this answers is not in the document]
Comment (nfelt14) [reply docs/advanced/signal_generators.md:55]: The overview page has the arrow pointing to the parent class. This has the arrow pointing to the child class.

[the remark this answers is not in the document]
Comment (nfelt14) [reply docs/advanced/signal_generators.md:55]: @qthompso, I think that the arrow style just needs to change. It needs to show composition, not inheritance.

[the remark this answers is not in the document]
Comment (qthompso) [reply docs/advanced/signal_generators.md:55]: It's arrow is pointing towards the parent class?

**Review by nfelt14 — DISMISSED** [review]

**Review by nfelt14 — APPROVED** [review]

File: src/tm_devices/helpers/constants_and_dataclasses.py [hunk 14]
```diff
@@ -704,4 +704,25 @@ def __str__(self) -> str:
         ),
     }
 )
+
+LOAD_IMPEDANCE_LOOKUP: Final[Mapping[Union[float, str], LoadImpedanceAFG]] = MappingProxyType(
```
Comment (nfelt14) [inline src/tm_devices/helpers/constants_and_dataclasses.py:708]: Why is this here twice?

File: CHANGELOG.md
Comment (nfelt14) [inline CHANGELOG.md:1]: The changes in this file seem a little off.

File: CHANGELOG.md [hunk 15]
```diff
@@ -192,15 +216,27 @@ Things to be included in the next release go here.
 
 ### Added
 
-- Added TekVISA as one of the VISA backends supported.
```
Comment (nfelt14) [inline CHANGELOG.md:195]: Stuff below this probably shouldn't have changed.

**Review by nfelt14 — CHANGES_REQUESTED** [review]

**Review by nfelt14 — APPROVED** [review]

**Review by ldantek — DISMISSED** [review]

**Review by nfelt14 — COMMENTED** [review]
Wait for a few more expert reviews

File: docs/advanced/signal_generators.md
Comment (nfelt14) [inline docs/advanced/signal_generators.md:1]: Let's wait for a few more expert reviews on this before merging

**Review by nfelt14 — DISMISSED** [review]

File: src/tm_devices/drivers/pi/signal_generators/afgs/afg.py [hunk 16]
```diff
@@ -0,0 +1,484 @@
...
+            offset_range=offset_range,
+            sample_rate_range=sample_rate_range,
+        )
+
+    ################################################################################################
+    # Private Methods
+    ################################################################################################
+    def _reboot(self) -> None:
+        """Reboot the device."""
+        # TODO: implement
```
Comment (ldantek) [inline src/tm_devices/drivers/pi/signal_generators/afgs/afg.py:229]: should these todo's be left in here? or should they raise the NotImplementedError? or they could be removed if the children of this class actually implement it?

File: src/tm_devices/drivers/pi/scopes/tekscope/tekscope.py [hunk 17]
```diff
@@ -379,61 +383,208 @@ def curve_query(  # noqa: PLR0912,C901
...
+            offset: The offset of the signal to generate.
+            burst_count: The number of wavelengths to be generated.
+            channel: The channel number to output the signal from, or 'all'.
+            output_signal_path: The output signal path of the specified channel.
+            termination: The impedance to set the channel to.
+            duty_cycle: The duty cycle to set the signal to.
+            polarity: The polarity to set the signal to.
+            symmetry: The symmetry to set the signal to, only applicable to certain functions.
+        """
+        del polarity, channel, output_signal_path  # these aren't used
```
Comment (ldantek) [inline src/tm_devices/drivers/pi/scopes/tekscope/tekscope.py:460]: should the docstring be edited to indicate which fields are ignored like this?

File: src/tm_devices/drivers/pi/signal_generators/awgs/awg.py [hunk 18]
```diff
@@ -0,0 +1,494 @@
...
+        self,
+        output_signal_path: Optional[SignalGeneratorOutputPathsBase],
+    ) -> Tuple[ParameterBounds, ParameterBounds, ParameterBounds]:
+        raise NotImplementedError
+
+    def _reboot(self) -> None:
+        """Reboot the device."""
+        # TODO: overwrite the reboot code here
+
+    # TODO: add testing for this
```
Comment (ldantek) [inline src/tm_devices/drivers/pi/signal_generators/awgs/awg.py:360]: todo still needed?

File: src/tm_devices/drivers/pi/signal_generators/awgs/awg.py [hunk 19]
```diff
@@ -0,0 +1,494 @@
...
+    @abstractmethod
+    def _get_series_specific_constraints(
+        self,
+        output_signal_path: Optional[SignalGeneratorOutputPathsBase],
+    ) -> Tuple[ParameterBounds, ParameterBounds, ParameterBounds]:
+        raise NotImplementedError
+
+    def _reboot(self) -> None:
+        """Reboot the device."""
+        # TODO: overwrite the reboot code here
```
Comment (ldantek) [inline src/tm_devices/drivers/pi/signal_generators/awgs/awg.py:358]: I think this can just be removed if the children re-define it

File: src/tm_devices/drivers/pi/signal_generators/awgs/awg.py [hunk 20]
```diff
@@ -0,0 +1,494 @@
...
+        elif value:
+            # No error is raised when 0 is the offset value and the output signal path is
+            # in a state where offset cannot be set.
+            offset_error = (
+                f"The offset can only be set with an output signal path of "
+                f"{self._awg.OutputSignalPath.DCA.value}."
+            )
+            raise ValueError(offset_error)
+
+    def set_state(self, value: int) -> None:
```
Comment (ldantek) [inline src/tm_devices/drivers/pi/signal_generators/awgs/awg.py:467]: Any value in making this just accept a bool to do away with the type checking?

**Review by ldantek — APPROVED** [review]

File: src/tm_devices/drivers/pi/signal_generators/awgs/awg.py (hunk 20 above)
Comment (qthompso) [reply src/tm_devices/drivers/pi/signal_generators/awgs/awg.py:467]: You can, but it ends up looking strange: set_state(True)

it could be changed to like, channel_on, but that doesn't really reflect the scpi command, considering it's a 1 to 1 method

File: src/tm_devices/drivers/pi/pi_device.py [hunk 21]
```diff
@@ -633,6 +706,70 @@ def set_and_check(
...
+            opc: Boolean indicating if ``*OPC?`` should be queried after sending the command.
+            allow_empty: Set this to True if an empty return string is permitted.
+            verify_value: Boolean indicating to verify value after write.
+
+        Returns:
+            Tuple containing the boolean value indicating if the command needed to be set and
+            the value returned from the query.
+        """
+        try:
+            query_passed, actual_value = self.query_response(
```
Comment (michaelwagoner) [inline src/tm_devices/drivers/pi/pi_device.py:746]: Potential nit but instead of using a try catch with query response would it make sense to just query the value instead and compare it to the actual value

**Review by michaelwagoner — DISMISSED** [review]

File: src/tm_devices/drivers/pi/pi_device.py (hunk 21 above)
Comment (qthompso) [reply src/tm_devices/drivers/pi/pi_device.py:746]: Query response does this inherently, I believe it makes more sense to just utilize already existing functionality

File: CHANGELOG.md [hunk 22]
```diff
@@ -16,7 +16,55 @@ Valid subsections within a version are:
 
 ## Unreleased
 
-Things to be included in the next release go here.
+- Things to be included in the next release go here.
```
Comment (nfelt14) [inline CHANGELOG.md:19]:
```suggestion
Things to be included in the next release go here.
```

File: CHANGELOG.md [hunk 23]
```diff
@@ -16,7 +16,55 @@ Valid subsections within a version are:
...
+
+- <span style="color:red">BREAKING CHANGE</span>. Changed the term "signal source" to "signal generator".
+    - All uses of this term are changed. Import paths now use signal_generator instead of signal_source.
+- <span style="color:red">BREAKING CHANGE</span>. Changed the function name of `generate_waveform()` to `generate_function()`.
+    - `generate_waveform()` only exists on AWGs now, however the functionality is entirely changed.
+- <span style="color:red">BREAKING CHANGE</span>. Changed the `generate_function()` function by removing burst functionality.
+    - Any use of burst now must use `setup_burst()` and `generate_burst()` instead.
+- Updated AWG's such that the `family_base_class` is at the series level.
+
+### Added
```
Comment (nfelt14) [inline CHANGELOG.md:45]: Duplicate sections

File: CHANGELOG.md [hunk 24]
```diff
@@ -192,15 +240,27 @@ Things to be included in the next release go here.
 
 ### Added
 
-- Added TekVISA as one of the VISA backends supported.
+- Added the constraint ranges for all signal generators
```
Comment (nfelt14) [inline CHANGELOG.md:243]: Why did this change? I already fixed this once.

File: src/tm_devices/drivers/pi/signal_generators/afgs/afg.py [hunk 25]
```diff
@@ -0,0 +1,472 @@
...
+        Args:
+            function: The function that needs to be generated.
+            waveform_length: The length of the waveform if no function or arbitrary is provided.
+            frequency: The frequency of the waveform that needs to be generated.
+            load_impedance: The suggested impedance on the source.
+
+        Returns:
+            Ranges for amplitude, frequency, offset, and sample rate.
+        """
+        raise NotImplementedError
```
Comment (nfelt14) [inline src/tm_devices/drivers/pi/signal_generators/afgs/afg.py:246]: Not needed for an abstract method.

File: src/tm_devices/drivers/pi/signal_generators/afgs/afg3k.py [hunk 26]
```diff
@@ -0,0 +1,184 @@
...
+
+        sample_rate_range = ParameterBounds(lower=sample_rate, upper=sample_rate)
+
+        return amplitude_range, frequency_range, offset_range, sample_rate_range
+
+    def _reboot(self) -> None:
+        """Reboot the device."""
+        # TODO: implement
+        raise NotImplementedError(
+            f"``.reboot()`` is not yet implemented for the {self.__class__.__name__} driver"
```
Comment (nfelt14) [inline src/tm_devices/drivers/pi/signal_generators/afgs/afg3k.py:183]:
```suggestion
            f"``._reboot()`` is not yet implemented for the {self.__class__.__name__} driver"
```

File: src/tm_devices/drivers/pi/signal_generators/awgs/awg.py [hunk 27]
```diff
@@ -0,0 +1,461 @@
...
+            needed_sample_rate = 15000000.0
+
+        return predefined_name, needed_sample_rate
+
+    @abstractmethod
+    def _get_series_specific_constraints(
+        self,
+        output_signal_path: Optional[SignalGeneratorOutputPathsBase],
+    ) -> Tuple[ParameterBounds, ParameterBounds, ParameterBounds]:
+        raise NotImplementedError
```
Comment (nfelt14) [inline src/tm_devices/drivers/pi/signal_generators/awgs/awg.py:351]: Not needed for abstract methods

File: src/tm_devices/drivers/pi/signal_generators/awgs/awg.py [hunk 28]
```diff
@@ -0,0 +1,461 @@
...
+
+    def _reboot(self) -> None:
+        """Reboot the device."""
+        # TODO: overwrite the reboot code here
```
Comment (nfelt14) [inline src/tm_devices/drivers/pi/signal_generators/awgs/awg.py:355]:
```suggestion
        # TODO: overwrite the reboot code here or in a child class
```

File: src/tm_devices/helpers/constants_and_dataclasses.py [hunk 29]
```diff
@@ -704,4 +704,14 @@ def __str__(self) -> str:
         ),
     }
 )
-"""A mapping of model USBTMC info."""
```
Comment (nfelt14) [inline src/tm_devices/helpers/constants_and_dataclasses.py:707]: Why was this removed again?

File: src/tm_devices/drivers/pi/signal_generators/awgs/awg5200.py [hunk 30]
```diff
@@ -0,0 +1,342 @@
...
+            output_signal_path_error = (
+                f"{value.value} is an invalid output signal path for {self._awg.model}."
+            )
+            raise ValueError(output_signal_path_error)
+        self._awg.set_if_needed(f"OUTPUT{self.num}:PATH", value.value)
+
+    def _reboot(self) -> None:
+        """Reboot the device."""
+        # TODO: implement
+        raise NotImplementedError(
```
Comment (nfelt14) [inline src/tm_devices/drivers/pi/signal_generators/awgs/awg5200.py:340]: Why not move this to the parent class? No need to define this unimplemented method twice.

File: src/tm_devices/drivers/pi/scopes/tekscope/tekscope.py [hunk 31]
```diff
@@ -61,19 +70,9 @@ class TekProbeData:
...
-    name: str
-    """Name of the channel."""
-    probe: TekProbeData
-    """Details about channel setup and any connected probe."""
-    termination: Optional[Literal[50, 250_000, 1_000_000]] = None
-    """Defines impedance at channel port."""
-
-
 # pylint: disable=too-many-public-methods
+# TODO: Move signal generator mixin to different abstract class, so that tekscope software
```
Comment (nfelt14) [inline src/tm_devices/drivers/pi/scopes/tekscope/tekscope.py:74]:
```suggestion
# TODO: Move signal generator mixin to different abstract class, so that TekScopePC
```

**Review by nfelt14 — CHANGES_REQUESTED** [review]

File: src/tm_devices/helpers/constants_and_dataclasses.py (hunk 29 above)
Comment (qthompso) [reply src/tm_devices/helpers/constants_and_dataclasses.py:707]: Removed by accident
